### ISSUE TITLE
Proxies with wildcards don't work with parameters

### ISSUE THREAD

**Issue by JontyMC** [Azure/azure-functions-host#3593]
Given a proxies.json file like this:
```json
{
  "$schema": "http://json.schemastore.org/proxies",
  "proxies": {
    "api": {
      "matchCondition": {
        "route": "/api/{*restOfPath}"
      },
      "backendUri": "https://localhost/api/{restOfPath}"
    }
  }
}
```
Calling a function with an http trigger and a route with a parameter, such as "test/{id}", with a url like "test/123" results in an error:
```
System.Private.CoreLib: Exception while executing function: Test. Microsoft.Azure.WebJobs.Host: Exception binding parameter 'id'. Microsoft.Azure.WebJobs.Host: Binding data does not contain expected value 'id'.
```
This works with a v1 function app. Does not work with v2 function app (either 2.0.12115.0 or 2.0.12134.0).

Workaround is to create a proxy for each endpoint that has a binding parameter.

**Comment by JontyMC** [comment]
I'm surprised no one else has raised this issue. Am I doing something wrong?

**Comment by ikkentim** [comment]
I'm running into the same issue

**Comment by RicardoNiepel** [comment]
Some workaround in the meantime:

``` c#
private static async Task<RouteValueDictionary> GetRouteValuesAsync(HttpRequest req, ExecutionContext context)
{
    var routingFeature = req.HttpContext.Features.Get<IRoutingFeature>();
    var webJobsRouter = routingFeature.RouteData.Routers.Single(r => r is WebJobsRouter) as WebJobsRouter;
    var functionRouteTemplate = webJobsRouter.GetFunctionRouteTemplate(context.FunctionName);

    var functionRoute = new Route(webJobsRouter, functionRouteTemplate, webJobsRouter.ConstraintResolver);
    var newRouteContext = new RouteContext(req.HttpContext);
    await functionRoute.RouteAsync(newRouteContext);
    return newRouteContext.RouteData.Values;
}

[FunctionName("UsersHttpTrigger")]
public static async Task<IActionResult> Run(
    [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "users/{id}")] HttpRequest req,
    ILogger log,
    ExecutionContext context)
{
    var routeValues = await GetRouteValuesAsync(req, context);
    var idValue = routeValues["id"];
    ...
}
```

**Comment by vfedonkin** [comment]
Please look at this setting:
https://review.docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#azurefunctionproxybackendurldecodeslashes 
The idea is that {restOfPath} will contain encode slashes in "https://localhost/api/{restOfPath}" unless this parameter is set to true.

**Comment by RicardoNiepel** [comment]
Doesn't help.

With `"AZURE_FUNCTION_PROXY_BACKEND_URL_DECODE_SLASHES": "true"` the issue is still there:

```
Executing 'Functions.api' (Reason='This function was programmatically called via the host APIs.', Id=47dd0320-b07d-442e-9e8d-017c98424a1a)
Executing request via Azure Function Proxies
Executing 'UsersHttpTrigger' (Reason='This function was programmatically called via the host APIs.', Id=c6fc8c25-5b3c-4bfd-b736-b101951afd24)
Executed 'UsersHttpTrigger' (Failed, Id=c6fc8c25-5b3c-4bfd-b736-b101951afd24)

System.Private.CoreLib: Exception while executing function: UsersHttpTrigger. Microsoft.Azure.WebJobs.Host:
  Exception binding parameter 'id'. Microsoft.Azure.WebJobs.Host: Binding data does not contain expected value 'id'.
```

**Comment by alexkarcher-msft** [comment]
I was able to repo this issue on my own with a JS function, and setting the app setting [AZURE_FUNCTION_PROXY_DISABLE_LOCAL_CALL](https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#azurefunctionproxydisablelocalcall) to true restored the correct behavior.

This looks like an issue with our component to route local calls directly to functions in the same app, and we're going to take a look at the root cause. In the meantime, that app setting should mitigate the issue.

**Comment by JontyMC** [comment]
AZURE_FUNCTION_PROXY_DISABLE_LOCAL_CALL = true does not work for me when running C# functions via visual studio. I have not tested other scenarios. In fact, setting that value breaks the work around I am using, which is to explicitly create a proxy for each endpoint that has a binding parameter. Tested today using:

Azure Functions Core Tools (2.4.419 Commit hash: c9c1724d002bd90b2e6b41393915ea3a26bcf0ce)
Function Runtime Version: 2.0.12332.0

**Comment by asselin** [comment]
AZURE_FUNCTION_PROXY_DISABLE_LOCAL_CALL = true does not work for me with node.js functions either.  If I enable that, the functions start returning 500s for all calls.

**Comment by alexkarcher-msft** [comment]
Ah, good to point out that with that application setting you'll need to actually put in your hostname instead of "localhost"

Locally you can still use localhost, but in Azure you'll need to use your domain name. I'd set it with an app setting if you're testing locally, so you can use different values each place.

**Comment by JontyMC** [comment]
I cannot get this to work using AZURE_FUNCTION_PROXY_DISABLE_LOCAL_CALL. Trying locally using Visual Studio and C# function project. With:

`"backendUri": "https://localhost/api/{restOfPath}"`

It gives 500's. When I use:

`"backendUri": "http://localhost:7071/api/{restOfPath}"`

It seems to go into an infinite loop.

**Comment by alexkarcher-msft** [comment]
It's because you have defined an infinite loop 🙂
```
"matchCondition": {
        "route": "/api/{*restOfPath}"
      },
      "backendUri": "https://localhost/api/{restOfPath}"
```

To get this workaround to work you'll either need non-overlapping routes with {*restofpath} or you'll need to put your proxies in a seperate app.

Putting your proxies on /frontend and your functions on /api will do the trick for testing.

This app setting sends requests back to the HTTP front end of the Function App, which causes the proxies to be re-evaluated.

**Comment by JontyMC** [comment]
Doh! Makes sense, thanks

**Comment by sm1th** [comment]
Having the same issue. We require `"AZURE_FUNCTION_PROXY_DISABLE_LOCAL_CALL` to be `false`. Tried `"AZURE_FUNCTION_PROXY_BACKEND_URL_DECODE_SLASHES": true` to no avail.

**Comment by JontyMC** [comment]
This has been open for a year now. Is no one using proxies or is everyone just using this workaround?

**Comment by UliPlabst** [comment]
Is there any progress on this? I'm still hitting this bug!

**Comment by mikkark** [comment]
I noticed this today. I will be resorting to the workaround provided by Ricardo almost 1,5 years ago. Would be great to have this fixed (currently marked as bug by Microsoft, after being triaged).

**Comment by fabiocav** [comment]
Adding this to triage for proper assignment.

@jeffhollan can you confirm the `P1` here?

**Comment by GrantHoffman** [comment]
Yeah I'm going to hop on the complain train.  I would like this fixed as well

**Comment by IanKemp** [comment]
Typical Microsoft:

> marks bug as P1  
> still hasn't fixed it over a year later

**Comment by YvesCandel** [comment]
Any workaround available when using HttpRequestMessage instead of HttpRequest? Like @RicardoNiepel his workaround...

**Comment by HavyRhl** [comment]
same problem.
Workarounds not viable for us. pls fix.

**Comment by drekuru** [comment]
omg same bug, its almost been 3 years.

**Comment by mtycholaz** [comment]
Spent a full 3 hours trying to figure this out, then remembered that I had visited this thread over two years ago. Please get this fixed!

**Comment by YvesCandel** [comment]
We have relieved ourselves of this issue by using Azure Static Web Apps for our front-ends and removing the proxies.json because we no longer need it now. However, using proxies for Azure Functions still remains really broken. Not only did it break these paths, it also broke all the logging of execution of our functions to the Azure Portal.

**Comment by saikatguha** [comment]
Any updates on this bug. Just spent 3 hours debugging.
Given the 2018 post date, this bug is old enough to go to Kindergarten!

**Comment by mikkok-adafy** [comment]
Azure Functions Proxies are no longer supported with Functions runtime 4.x. Takes care of this issue, more or less.
